crates/but/src/id/tests.rs

```diff
@@ -0,0 +1,184 @@
+        let actual = id_map.parse_str(&commit_id.to_hex_with_len(2).to_string())?;
+        assert_eq!(actual.len(), 1, "The commit can be resolved");
+        assert!(
+            matches!(&actual[0], CliId::Commit { oid } if oid == commit_id,),
+            "The id refers to the right commit"
+        );
+    }
+    Ok(())
+}
+
+#[test]
+fn unassigned_area_id_is_unambiguous() -> anyhow::Result<()> {
+    let stacks = &[stack([segment("branch001", [id(1)], None, [])])];
+    let id_map = IdMap::new(stacks)?;
+
+    assert_eq!(
+        id_map.unassigned().to_string(),
+        "000",
+        "the ID of the unassigned area should have enough 0s to be unambiguous"
+    );
+    Ok(())
+}
+
+#[test]
+fn non_commit_ids_do_not_collide() -> anyhow::Result<()> {
+    let stacks = &[stack([segment("h0", [id(2)], Some(id(1)), [])])];
+    let mut id_map = IdMap::new(stacks)?;
+    let changed_paths_fn = |commit_id: gix::ObjectId,
+                            parent_id: Option<gix::ObjectId>|
+     -> anyhow::Result<Vec<BString>> {
+        Ok(if commit_id == id(2) && parent_id == Some(id(1)) {
+            vec![
+                BString::from(b"committed1.txt"),
+                BString::from(b"committed2.txt"),
+            ]
+        } else {
+            bail!("unexpected IDs {} {:?}", commit_id, parent_id);
+        })
+    };
+    let hunk_assignments = vec![
+        hunk_assignment("uncommitted1.txt", None),
+        hunk_assignment("uncommitted2.txt", None),
+    ];
+    id_map.add_file_info(changed_paths_fn, hunk_assignments)?;
+
+    assert!(
+        matches!(
+        id_map.parse_str("g0")?.as_slice(),
+        [CliId::UncommittedFile{path,..}] if path == b"uncommitted1.txt"),
+        "Uncommitted files come first"
+    );
+    assert!(
+        matches!(
+        id_map.parse_str("h0")?.as_slice(),
+        [CliId::Branch{name,..}] if name == "h0"),
+        "uncommitted files do not collide with branches"
+    );
+    assert!(
+        matches!(
+        id_map.parse_str("i0")?.as_slice(),
+        [CliId::UncommittedFile{path,..}] if path == b"uncommitted2.txt"),
+        "uncommitted files also don't collide with themselves"
+    );
+    assert!(
+        matches!(
+        id_map.parse_str("j0")?.as_slice(),
+        [CliId::CommittedFile{commit_oid, path,..}] if *commit_oid == id(2) && path == b"committed1.txt"),
+        "then come committed files, as per incremented prefix"
+    );
+    assert!(
+        matches!(
+        id_map.parse_str("k0")?.as_slice(),
+        [CliId::CommittedFile{commit_oid, path,..}] if *commit_oid == id(2) && path == b"committed2.txt"),
+        "committed files also don't collide with themselves"
+    );
+    Ok(())
+}
+
+mod util {
+    use bstr::BString;
+    use but_core::ref_metadata::StackId;
+    use but_hunk_assignment::HunkAssignment;
+    use but_workspace::branch::Stack;
+    use but_workspace::ref_info::{Commit, LocalCommit, Segment};
```

Comment:
I put the utilities into their own module primarily to isolate their dependencies. Additionally, having the modules makes clear what is supposed to be used by tests, while allowing internal private functions.